File: ts/packages/trellis/auth/authorization/refresh.ts
import { Value } from "typebox/value";
import { ulid } from "ulid";

import { decodeTrellisHttpError, TrellisHttpError } from "../http_error.ts";
import { recordCatalogCounter } from "../../telemetry/metrics.ts";
import type { TrellisAuth } from "../session_auth.ts";
import type { AuthorizationContextCache } from "./client_context.ts";
import type {
  AuthorizationContextRefreshResponse,
  AuthorizationContextRefreshResult,
  AuthorizationRuntimeBinding,
  VerifiedAuthorizationContext,
} from "./types.ts";
import { AuthorizationContextRefreshResponseSchema as ResponseSchema } from "./types.ts";

/** HTTP refresh failure with terminal-state classification. */
export class AuthorizationContextRefreshError extends TrellisHttpError {
  readonly terminal: boolean;
  readonly loginInvalid: boolean;

  constructor(status: number, code: string) {
    super(status, code);
    this.name = "AuthorizationContextRefreshError";
    this.terminal = [
      "session_not_found",
      "session_expired",
      "session_revoked",
      "identity_not_found",
      "identity_inactive",
      "user_not_found",
      "user_inactive",
      "login_not_found",
      "participant_not_found",
      "participant_changed",
      "contract_changed",
      "authority_not_found",
      "authority_rejected",
      "authority_revoked",
      "authority_expired",
      "context_owner_mismatch",
      "deployment_inactive",
      "instance_inactive",
      "device_inactive",
      "activation_required",
      "delegation_expired",
      "context_refresh_mismatch",
      "invalid_proof",
    ].includes(code);
    this.loginInvalid = [
      "session_not_found",
      "session_expired",
      "session_revoked",
      "identity_not_found",
      "identity_inactive",
      "user_not_found",
      "user_inactive",
      "login_not_found",
    ].includes(code);
  }
}

/** Refresh a context after proving possession of its bound session key. */
export async function refreshAuthorizationContextWithMetadata(args: {
  trellisUrl: string;
  sessionId: string;
  auth: TrellisAuth;
  sessionKey?: string;
  cache: AuthorizationContextCache;
  fetch?: typeof globalThis.fetch;
  shouldInstall?: () => boolean;
  requiredTransport?: "native" | "websocket";
  prepareInstall?: (
    response: AuthorizationContextRefreshResponse,
  ) => Promise<(verified: VerifiedAuthorizationContext) => void>;
  prepareOnly?: boolean;
}): Promise<AuthorizationContextRefreshResult> {
  const fetch = args.fetch ?? globalThis.fetch;
  const currentDigest = args.cache.storedContextDigest();
  let runtime: AuthorizationRuntimeBinding | undefined;
  try {
    runtime = args.cache.runtimeBinding();
  } catch {
    runtime = undefined;
  }
  if (runtime?.loginSessionId && runtime.loginSessionId !== args.sessionId) {
    throw new Error("authorization recovery session mismatch");
  }
  const requestStartedAt = args.cache.nowMilliseconds();
  const unsignedRequest = {
    requestId: ulid(),
    issuedAt: Math.trunc(args.auth.currentIat() * 1_000),
    loginSessionId: args.sessionId,
    connectionId: runtime?.connectionId ?? ulid(),
    sessionKey: args.sessionKey ?? args.auth.sessionKey,
    currentContextDigest: currentDigest ?? null,
  };
  const proof = await args.auth.signSessionProof({
    purpose: "authorizationContextRefresh",
    origin: new URL(args.trellisUrl).origin,
    sessionPublicKey: args.auth.sessionKey,
    unsignedRequest,
  });
  let outcome = "error";
  let response: Response;
  try {
    response = await fetch(
      new URL("/auth/context/refresh", args.trellisUrl),
      {
        method: "POST",
        headers: { "content-type": "application/json" },
        body: JSON.stringify({ ...unsignedRequest, proof }),
      },
    );
    if (!response.ok) {
      const error = await decodeTrellisHttpError(response);
      const classified = new AuthorizationContextRefreshError(
        error.status,
        error.code,
      );
      outcome = classified.terminal
        ? "terminal"
        : error.code === "resource_pending"
        ? "pending"
        : response.status === 503
        ? "unavailable"
        : "error";
      throw classified;
    }
    outcome = "ok";
  } catch (error) {
    if (error instanceof DOMException && error.name === "AbortError") {
      outcome = "cancelled";
    }
    throw error;
  } finally {
    recordCatalogCounter("trellis.auth.refresh.attempts", 1, {
      "trellis.participant.kind": "user",
      "trellis.outcome": outcome,
    });
  }
  const next = Value.Parse(
    ResponseSchema,
    await response.json(),
  ) as AuthorizationContextRefreshResponse;
  if (
    args.requiredTransport &&
    !next.transports[args.requiredTransport]?.natsServers.length
  ) {
    throw new Error(
      `authorization refresh has no ${args.requiredTransport} NATS endpoints`,
    );
  }
  const serverClockOffsetMs = next.serverNow - Math.trunc(
    (requestStartedAt + args.cache.nowMilliseconds()) / 2,
  );
  args.cache.setServerClockOffsetMs(serverClockOffsetMs);
  args.auth.setServerClockOffsetMs(serverClockOffsetMs);
  if (args.shouldInstall?.() === false) {
    throw new Error("authorization context refresh stopped");
  }
  const nextRuntime = runtimeBindingFromResponse(next);
  const context = await args.cache[args.prepareOnly ? "prepare" : "install"](
    next.authorizationContext,
    {
      bootstrapJwt: next.routing.bootstrapJwt,
      bootstrapJwtExpiresAt: next.routing.bootstrapJwtExpiresAt,
    },
    Math.floor(next.serverNow / 1_000),
    args.shouldInstall,
    nextRuntime,
    async (verified) => {
      const installAdditional = await args.prepareInstall?.(next);
      return installAdditional ? () => installAdditional(verified) : undefined;
    },
  );
  return { context, response: next };
}

/** Refresh a context and return only its verified projection. */
export async function refreshAuthorizationContext(args: {
  trellisUrl: string;
  sessionId: string;
  auth: TrellisAuth;
  sessionKey?: string;
  cache: AuthorizationContextCache;
  fetch?: typeof globalThis.fetch;
  shouldInstall?: () => boolean;
}): Promise<VerifiedAuthorizationContext> {
  const result = await refreshAuthorizationContextWithMetadata(args);
  return result.context;
}

/** Start proactive refresh using the context's distributed refresh time. */
export function startAuthorizationContextRefresh(args: {
  trellisUrl: string;
  sessionId: string;
  auth: TrellisAuth;
  sessionKey?: string;
  cache: AuthorizationContextCache;
  fetch?: typeof globalThis.fetch;
  refresh?: (
    shouldInstall: () => boolean,
  ) => Promise<VerifiedAuthorizationContext>;
  onTerminalFailure?: (error: unknown) => void | Promise<void>;
  onTransientFailure?: (error: unknown) => void | Promise<void>;
  onRefresh?: (context: VerifiedAuthorizationContext) => void | Promise<void>;
}): () => void {
  let stopped = false;
  let timer: ReturnType<typeof setTimeout> | undefined;
  let failures = 0;
  let running = false;
  let wakePending = false;
  const schedule = (delayMs: number) => {
    if (!stopped) timer = setTimeout(run, delayMs);
  };
  const run = async () => {
    if (running) {
      wakePending = true;
      return;
    }
    running = true;
    const clearGuard = args.cache.clearGuard();
    try {
      let before: string | undefined;
      try {
        before = args.cache.storedContextDigest();
      } catch {
        before = undefined;
      }
      const context = args.refresh
        ? await args.refresh(() => !stopped)
        : (await refreshAuthorizationContextWithMetadata({
          ...args,
          shouldInstall: () => !stopped,
          prepareOnly: true,
        })).context;
      if (stopped) return;
      failures = 0;
      await args.onRefresh?.(context);
      schedule(
        refreshDelay(
          args.cache,
          before === context.contextDigest ? 5_000 : 1_000,
        ),
      );
    } catch (error) {
      if (stopped) return;
      if (error instanceof AuthorizationContextRefreshError && error.terminal) {
        if (!(await args.cache.clearIfCurrent(clearGuard))) {
          failures = 0;
          schedule(refreshDelay(args.cache, 1_000));
          return;
        }
        await args.onTerminalFailure?.(error);
        return;
      }
      failures += 1;
      let current: VerifiedAuthorizationContext | undefined;
      try {
        current = args.cache.current();
      } catch {
        current = undefined;
      }
      await args.onTransientFailure?.(error);
      const beforeExpiry = current
        ? Math.max(
          1_000,
          (current.context.expiresAt - args.cache.correctedNowSeconds()) *
            1_000,
        )
        : Number.POSITIVE_INFINITY;
      schedule(Math.min(beforeExpiry, 5_000 * 2 ** Math.min(failures - 1, 3)));
    } finally {
      running = false;
      if (wakePending && !stopped) {
        wakePending = false;
        if (timer !== undefined) clearTimeout(timer);
        schedule(0);
      }
    }
  };
  schedule(refreshDelay(args.cache));
  let unregisterRefreshRequest: () => void;
  try {
    unregisterRefreshRequest = args.cache.registerRefreshRequest(() => {
      if (running) {
        wakePending = true;
        return;
      }
      if (timer !== undefined) clearTimeout(timer);
      schedule(0);
    });
  } catch (error) {
    stopped = true;
    if (timer !== undefined) clearTimeout(timer);
    throw error;
  }
  return () => {
    stopped = true;
    unregisterRefreshRequest();
    if (timer !== undefined) clearTimeout(timer);
  };
}

function refreshDelay(
  cache: AuthorizationContextCache,
  minimumMs = 1_000,
): number {
  try {
    return Math.max(
      minimumMs,
      (cache.routingRefreshAt() - cache.correctedNowSeconds()) * 1_000,
    );
  } catch {
    return minimumMs;
  }
}

function runtimeBindingFromResponse(
  response: AuthorizationContextRefreshResponse,
): AuthorizationRuntimeBinding {
  if (
    !response.transports.native &&
    !response.transports.websocket
  ) {
    throw new Error("authorization refresh returned no NATS transport");
  }
  return {
    connectionId: response.runtime.connectionId,
    loginSessionId: response.runtime.loginSessionId,
    participantId: response.runtime.participantId,
    inboxPrefix: response.runtime.inboxPrefix,
    transports: {
      ...(response.transports.native === undefined ? {} : {
        native: {
          natsServers: [...response.transports.native.natsServers],
        },
      }),
      ...(response.transports.websocket === undefined ? {} : {
        websocket: {
          natsServers: [...response.transports.websocket.natsServers],
        },
      }),
    },
  };
}

File: crates/trellis/src/client/authorization/refresh.rs
use std::sync::Arc;
use std::time::Duration;

use serde_json::json;
use trellis_protocol::{
    AuthorizationContextRefreshSessionProofInput, AuthorizationPrincipalKind,
    NativeBootstrapSessionProofInput, SessionProofInput,
};

use super::super::connection::{
    apply_native_authorization_refresh, AppliedNativeAuthorization,
    NativeAuthorizationRefreshContext,
};
use super::super::{proof::new_request_id, SessionAuth, TrellisClientError};
use super::own_context::{
    system_now_millis, AuthorizationContextCache, AuthorizationRefreshRequest,
};
use super::provider_cache::AuthorizationProviderCache;
use super::rotation::AuthorizationTransportRotation;
use super::types::{AuthorizationCredential, AuthorizationInstallation};

/// Authorization codes that report in-flight materialization rather than denial.
///
/// The server returns these while approved authority, dependencies, or resources
/// are still being materialized. They are retriable with a bounded backoff: a
/// participant's growing authority must never fail a caller.
pub(crate) fn is_retriable_authorization_code(code: &str) -> bool {
    matches!(
        code,
        "resource_pending" | "dependency_pending" | "authorization_pending"
    )
}

fn is_terminal_refresh_error(code: &str) -> bool {
    matches!(
        code,
        "session_not_found"
            | "session_expired"
            | "session_revoked"
            | "user_not_found"
            | "user_inactive"
            | "identity_not_found"
            | "identity_revoked"
            | "grant_binding_missing"
            | "grant_binding_revoked"
            | "grant_binding_expired"
            | "participant_not_installed"
            | "deployment_inactive"
            | "instance_inactive"
            | "device_inactive"
            | "activation_required"
            | "login_not_found"
            | "context_owner_mismatch"
            | "authority_revoked"
            | "authority_expired"
            | "authority_not_found"
            | "delegation_expired"
            | "context_refresh_mismatch"
            | "invalid_proof"
    )
}

/// Stable, connection-owned collaborators for authorization-context refresh.
///
/// One connection owns exactly this set; grouping them keeps the physical
/// refresh transaction and the background refresh task from threading the same
/// eight independent parameters and makes the owned surface explicit.
pub(crate) struct AuthorizationRefreshRuntime {
    pub(crate) contexts: Arc<AuthorizationContextCache>,
    pub(crate) auth: Arc<SessionAuth>,
    pub(crate) nats: async_nats::Client,
    pub(crate) applied_native_authorization: Arc<tokio::sync::Mutex<AppliedNativeAuthorization>>,
    pub(crate) provider: AuthorizationProviderCache,
    pub(crate) rotation: Arc<AuthorizationTransportRotation>,
    pub(crate) live: Option<Arc<crate::live::manager::LiveSessionManager>>,
    pub(crate) timeout_ms: u64,
}

/// Obtain or renew connection authority using only the owner credential and proof.
pub(crate) async fn refresh(
    cache: &AuthorizationContextCache,
    auth: &SessionAuth,
    promote: bool,
) -> Result<(String, bool), TrellisClientError> {
    if auth.session_key != cache.session_key {
        return Err(TrellisClientError::Bootstrap(
            "refresh signing key does not belong to this connection".into(),
        ));
    }
    let observed_digest = cache.stored_context_digest().ok();
    let _refresh = cache.lock_refresh().await;
    if observed_digest != cache.stored_context_digest().ok() {
        return Ok((cache.stored_context_digest()?, false));
    }
    let previous = cache.state_snapshot()?;
    let request_started_at = system_now_millis()?;
    let issued_at = cache.corrected_now_millis()?;
    let mut request = json!({"requestId": new_request_id(), "connectionId": cache.connection_id});
    if let Some(name) = &cache.name {
        request["name"] = json!(name);
    }
    let (route, input, signer) = match cache.credential.as_ref() {
        AuthorizationCredential::Native {
            kind,
            identity,
            package_evidence,
            participant_path,
            companion,
        } => {
            request["identityKeyId"] = json!(identity.key_id());
            request["sessionKey"] = json!(auth.session_key);
            request["iat"] = json!(issued_at);
            request["packageEvidence"] = serde_json::to_value(package_evidence)?;
            request["participantPath"] = json!(participant_path);
            request["packageDigest"] = json!(package_evidence.root_digest());
            if let Some(companion) = companion {
                let companion_connection_id = ulid::Ulid::new().to_string();
                let companion_request_id = ulid::Ulid::new().to_string();
                let (_, companion_session_key) = crate::auth::generate_session_keypair();
                let digest = trellis_protocol::digest_json(&json!({
                    "format": "trellis.device.user-companion.v1",
                    "origin": cache.http().origin(),
                    "identityKeyId": identity.key_id(),
                    "participantId": companion.participant_id,
                    "connectionId": companion_connection_id,
                    "requestId": companion_request_id,
                    "issuedAt": issued_at,
                    "sessionKey": companion_session_key,
                }))
                .map_err(|error| TrellisClientError::Bootstrap(error.to_string()))?;
                let digest = crate::client::proof::base64url_decode(&digest)?;
                request["companion"] = json!({
                    "connectionId": companion_connection_id,
                    "requestId": companion_request_id,
                    "issuedAt": issued_at,
                    "sessionKey": companion_session_key,
                    "proof": companion.installation.sign_bytes(&digest),
                });
            }
            let input = NativeBootstrapSessionProofInput {
                origin: cache.http().origin(),
                unsigned_request: request.clone(),
            };
            match kind {
                AuthorizationPrincipalKind::Service => (
                    "/bootstrap/service",
                    SessionProofInput::service_bootstrap(input),
                    identity.as_ref(),
                ),
                AuthorizationPrincipalKind::Device => (
                    "/bootstrap/device",
                    SessionProofInput::device_bootstrap(input),
                    identity.as_ref(),
                ),
                AuthorizationPrincipalKind::User => {
                    return Err(TrellisClientError::Bootstrap(
                        "user login cannot use a native credential".into(),
                    ))
                }
            }
        }
        AuthorizationCredential::User {
            login_session_id,
            installation,
        } => {
            request["loginSessionId"] = json!(login_session_id);
            request["issuedAt"] = json!(issued_at);
            request["sessionKey"] = json!(cache.session_key);
            request["currentContextDigest"] = json!(observed_digest);
            (
                "/auth/context/refresh",
                SessionProofInput::authorization_context_refresh(
                    AuthorizationContextRefreshSessionProofInput {
                        origin: cache.http().origin(),
                        session_public_key: installation.session_key.clone(),
                        unsigned_request: request.clone(),
                    },
                ),
                installation.as_ref(),
            )
        }
    };
    let input = input.map_err(|error| TrellisClientError::Bootstrap(error.to_string()))?;
    request["proof"] = serde_json::to_value(signer.sign_session_proof(&input)?)?;
    let response = cache.http().post_json(route, &request).await?;
    let server_now = response["serverNow"]
        .as_i64()
        .filter(|now| (0..=9_007_199_254_740_991).contains(now))
        .ok_or_else(|| {
            TrellisClientError::Bootstrap("bootstrap omitted a safe server time".into())
        })?;
    let midpoint = request_started_at
        .checked_add(system_now_millis()?)
        .and_then(|sum| sum.checked_div(2))
        .ok_or_else(|| TrellisClientError::Bootstrap("bootstrap time overflow".into()))?;
    let mut runtime = response["runtime"].clone();
    runtime
        .as_object_mut()
        .ok_or_else(|| TrellisClientError::Bootstrap("bootstrap runtime is not an object".into()))?
        .insert("transports".into(), response["transports"].clone());
    let mut authorization = response
        .get("authorization")
        .filter(|value| !value.is_null())
        .cloned();
    if let (Some(object), Some(companion)) = (
        authorization
            .as_mut()
            .and_then(serde_json::Value::as_object_mut),
        response.get("companion").filter(|value| !value.is_null()),
    ) {
        object.insert("companion".to_owned(), companion.clone());
    }
    let installation = AuthorizationInstallation {
        context: serde_json::from_value(response["authorizationContext"].clone())?,
        routing: serde_json::from_value(response["routing"].clone())?,
        runtime: serde_json::from_value(runtime)?,
        api_bindings: serde_json::from_value(response["apiBindings"].clone())?,
        server_clock_offset_ms: server_now
            .checked_sub(midpoint)
            .ok_or_else(|| TrellisClientError::Bootstrap("bootstrap time overflow".into()))?,
        authorization,
    };
    let context_digest = if promote {
        cache.install_initial(installation)?;
        cache.retained_context_digest()?
    } else {
        cache.prepare(installation)?
    };
    Ok((
        context_digest,
        previous.runtime.as_ref() != Some(&cache.runtime_binding()?),
    ))
}

/// Retry one signed bootstrap/context refresh until materialization settles.
///
/// The server reports a retriable authorization code while approved authority,
/// dependencies, or resources are still being materialized; a participant's
/// growing authority must never fail a caller. Only those codes are retried,
/// with bounded backoff, under a single overall budget equal to the caller's
/// connection timeout. Terminal authorization errors and non-retriable
/// infrastructure errors return immediately, and exhausting the budget
/// reports authorization as unavailable rather than looping forever.
pub(crate) async fn refresh_until_materialized(
    contexts: &AuthorizationContextCache,
    auth: &SessionAuth,
    timeout_ms: u64,
) -> Result<(), TrellisClientError> {
    let mut retry_delay = Duration::from_millis(100);
    tokio::time::timeout(Duration::from_millis(timeout_ms), async {
        loop {
            match contexts.refresh(auth).await {
                Err(TrellisClientError::BootstrapHttp { code, .. })
                    if is_retriable_authorization_code(&code) =>
                {
                    tokio::time::sleep(retry_delay).await;
                    retry_delay = (retry_delay * 2).min(Duration::from_secs(1));
                }
                result => {
                    result?;
                    break;
                }
            }
        }
        Ok::<(), TrellisClientError>(())
    })
    .await
    .map_err(|_| {
        TrellisClientError::AuthorizationUnavailable(
            "bootstrap resource materialization exceeded the connect budget".to_owned(),
        )
    })??;
    Ok(())
}

/// Install one prepared authorization candidate as transparent maintenance.
///
/// The physical NATS attachment may rotate to admit refreshed routing
/// credentials, but the logical connection is preserved. The predecessor
/// stays application-current until the candidate is admitted, its exact
/// revocation coverage is retained on the replacement attachment, and the
/// candidate is promoted. The planned-rotation marker is held across the whole
/// transaction so a failure never leaves a logically connected attachment that
/// is not actually usable.
pub(crate) async fn install_prepared_authorization(
    runtime: &AuthorizationRefreshRuntime,
) -> Result<String, TrellisClientError> {
    let (candidate_digest, _) = runtime.contexts.prepare_refresh(&runtime.auth).await?;
    let refreshed = AppliedNativeAuthorization::from_cache(&runtime.contexts)?;
    let mut applied = runtime.applied_native_authorization.lock().await;
    let rotates = applied.rotates_to(&refreshed);
    // Planned-rotation maintenance only applies while the physical attachment is
    // healthy. During an already-real outage the refreshed credential is still
    // installed, but the reconnect must take the ordinary recovery branch so the
    // connection and its Live manager are resumed.
    let planned =
        rotates && runtime.nats.connection_state() == async_nats::connection::State::Connected;
    if planned {
        runtime.provider.begin_planned_rotation();
    }
    let mut context = NativeAuthorizationRefreshContext {
        nats: &runtime.nats,
        applied: &mut applied,
        contexts: &runtime.contexts,
        rotation: &runtime.rotation,
        live: runtime.live.as_deref(),
        timeout_ms: runtime.timeout_ms,
    };
    if let Err(error) = apply_native_authorization_refresh(&mut context, refreshed, planned).await {
        runtime.provider.abandon_rotation();
        return Err(error);
    }
    if let Err(error) = runtime
        .provider
        .retain_own_context(&candidate_digest, runtime.provider.epoch())
        .await
    {
        runtime.provider.abandon_rotation();
        runtime.rotation.cancel();
        return Err(error);
    }
    if let Err(error) = runtime
        .provider
        .finalize_own_installation(&candidate_digest, true)
    {
        runtime.provider.abandon_rotation();
        runtime.rotation.cancel();
        return Err(error);
    }
    runtime.rotation.complete();
    Ok(candidate_digest)
}

/// Background own-context refresh on the retained NATS connection.
pub(crate) fn spawn_authorization_context_refresh_task(
    runtime: AuthorizationRefreshRuntime,
) -> tokio::task::JoinHandle<()> {
    tokio::spawn(async move {
        loop {
            let delay = match runtime.contexts.refresh_delay() {
                Ok(delay) => delay,
                Err(error) => {
                    tracing::warn!(%error, "authorization context refresh stopped");
                    return;
                }
            };
            let request = tokio::select! {
                () = tokio::time::sleep(delay) => AuthorizationRefreshRequest {
                    context_digest: runtime.contexts.stored_context_digest().ok(),
                    refresh_credential: true,
                },
                request = runtime.contexts.wait_refresh_request() => request,
            };
            if request.context_digest.is_some()
                && request.context_digest != runtime.contexts.stored_context_digest().ok()
            {
                continue;
            }
            if !request.refresh_credential && request.context_digest.is_some() {
                if let Some(digest) = request.context_digest {
                    let epoch = runtime.provider.epoch();
                    match runtime.provider.retain_own_context(&digest, epoch).await {
                        Ok(()) => {
                            let promote = runtime.contexts.candidate_digest().is_ok();
                            match runtime.provider.finalize_own_installation(&digest, promote) {
                                Ok(()) => continue,
                                Err(error) => {
                                    tracing::warn!(%error, "authorization coverage publication failed")
                                }
                            }
                        }
                        Err(error) => {
                            tracing::warn!(%error, "authorization coverage reconciliation will retry")
                        }
                    }
                }
                tokio::time::sleep(std::time::Duration::from_secs(5)).await;
                runtime.contexts.request_coverage_reconciliation();
                continue;
            }
            match install_prepared_authorization(&runtime).await {
                Ok(_) => {}
                Err(TrellisClientError::BootstrapHttp { status, code })
                    if is_terminal_refresh_error(&code) =>
                {
                    tracing::warn!(status, "authorization context refresh rejected");
                    if let Err(error) = runtime.contexts.clear() {
                        tracing::warn!(%error, "failed to clear rejected authorization context");
                    }
                    let _ = runtime.nats.drain().await;
                    return;
                }
                Err(error) => {
                    tracing::warn!(%error, "authorization context refresh will retry");
                    tokio::time::sleep(std::time::Duration::from_secs(5)).await;
                    runtime.contexts.request_refresh();
                }
            }
        }
    })
}

#[cfg(test)]
mod tests {
    use super::is_terminal_refresh_error;

    #[test]
    fn refresh_terminality_uses_exact_machine_codes() {
        assert!(is_terminal_refresh_error("session_revoked"));
        assert!(is_terminal_refresh_error("grant_binding_revoked"));
        assert!(is_terminal_refresh_error("context_refresh_mismatch"));
        assert!(is_terminal_refresh_error("login_not_found"));
        assert!(is_terminal_refresh_error("context_owner_mismatch"));
        assert!(!is_terminal_refresh_error("required_resources_unavailable"));
        assert!(!is_terminal_refresh_error("dependency_pending"));
        assert!(!is_terminal_refresh_error("resource_pending"));
        assert!(!is_terminal_refresh_error("authorization_pending"));
        assert!(!is_terminal_refresh_error("session_revoked later"));
    }
}
